### PR TITLE
Add Condition for aritrary SQL.

### PR DESCRIPTION
**Pull request by dseomn**
I think this will make it significantly easier to write new conditions,
including a STARTS_WITH() condition I want to add. I think it could also
simplify a bunch of the code in the existing conditions.

### CODE REVIEW

**Comment by dseomn** [comment]
@dgorelik If you think this is a good API, I'm curious what you think about things like the SQL `STARTS_WITH()` function: should I add a python function to this module specific to `STARTS_WITH()`, or should I use `ArbitraryCondition` directly in the code where I want to call `STARTS_WITH()`? I don't have much of an opinion either way. On the one hand, writing `condition.starts_with()` means a bit less code everywhere else that uses it; on the other hand, I'm not sure we should be in the habit of writing a new python function every time we want to use a new SQL function.

File: spanner_orm/tests/condition_test.py [hunk 1]
```diff
@@ -15,12 +15,132 @@
...
+          'FALSE',
+          (),
+      ),
+      (
+          'full',
+          condition.ArbitraryCondition(
+              '$key = IF($true_param, ${key_param}, $value_1)',
+              dict(
+                  key=models.SmallTestModel.key,
+                  true_param=condition.Param(
```
Comment (dgorelik) [inline spanner_orm/tests/condition_test.py:62]: Do you think we may want to try to infer the type from the python type? This looks a bit verbose to me, and I think condensing the number of line `condition.Param` takes up (enabled by allowing the type to be omitted) would help.

**Comment by dgorelik** [comment]
> @dgorelik If you think this is a good API, I'm curious what you think about things like the SQL `STARTS_WITH()` function: should I add a python function to this module specific to `STARTS_WITH()`, or should I use `ArbitraryCondition` directly in the code where I want to call `STARTS_WITH()`? I don't have much of an opinion either way. On the one hand, writing `condition.starts_with()` means a bit less code everywhere else that uses it; on the other hand, I'm not sure we should be in the habit of writing a new python function every time we want to use a new SQL function.

Good question. I don't really feel strongly either, but there seem to be quite a bit of conditions here already, so it seems there would be limited marginal maintenance cost of adding another one for `STARTS_WITH`.

File: spanner_orm/tests/condition_test.py (hunk 1 above)
Comment (dseomn) [reply spanner_orm/tests/condition_test.py:62]: I was thinking about that but decided to start with it being explicit. Happy to add inference in another PR. (The inference code ended up being a decent hunk of code.)

Comment (dgorelik) [reply spanner_orm/tests/condition_test.py:62]: Sounds good! Looking forward to a follow-up :)

**Review by dgorelik — APPROVED** [review]

**Comment by dseomn** [comment]
> > @dgorelik If you think this is a good API, I'm curious what you think about things like the SQL `STARTS_WITH()` function: should I add a python function to this module specific to `STARTS_WITH()`, or should I use `ArbitraryCondition` directly in the code where I want to call `STARTS_WITH()`? I don't have much of an opinion either way. On the one hand, writing `condition.starts_with()` means a bit less code everywhere else that uses it; on the other hand, I'm not sure we should be in the habit of writing a new python function every time we want to use a new SQL function.
> 
> Good question. I don't really feel strongly either, but there seem to be quite a bit of conditions here already, so it seems there would be limited marginal maintenance cost of adding another one for `STARTS_WITH`.

I think most of the ones here now are really common operations, compared to `STARTS_WITH`. And `contains()` has some escaping logic that's better done in a centralized place. I think I'm currently leaning towards only adding new conditions to this file if they would save a significant amount of code because of how common they are, or if they have some logic that's useful to centralize. I think I'd rather not start adding python functions for every ~one-liner use of `ArbitraryCondition` like `STARTS_WITH()`.

**Pull request merged**